lib/popcorn/pop_handlers.nit

```diff
@@ -0,0 +1,427 @@
+#
+# The following example creates a router as a module, loads a middleware handler in it,
+# defines some routes, and mounts the router module on a path in the main app.
+#
+# ~~~
+# class AppHome
+#	super Handler
+#
+#	redef fun get(req, res) do res.send "Site Home"
+# end
+#
+# class UserLogger
+#	super Handler
+#
+#	redef fun all(req, res) do print "User logged"
+# end
+#
+# class UserHome
+#	super Handler
+#
+#	redef fun get(req, res) do res.send "User Home"
+# end
+#
+# class UserProfile
+#	super Handler
+#
+#	redef fun get(req, res) do res.send "User Profile"
+# end
+#
+# var user_router = new Router
+# user_router.use("/*", new UserLogger)
+# user_router.use("/", new UserHome)
+# user_router.use("/profile", new UserProfile)
+#
+# var app = new App
+# app.use("/", new AppHome)
+# app.use("/user", user_router)
+# ~~~
+#
+# The app will now be able to handle requests to /user and /user/profile, as well
+# as call the `Time` middleware handler that is specific to the route.
+class Router
+	super Handler
+
+	# List of handlers to match with requests.
+	private var handlers = new Map[AppRoute, Handler]
+
+	# Register a `handler` for a route `path`.
+	#
+	# Route paths are matched in registration order.
+	fun use(path: String, handler: Handler) do
+		var route
+		if handler isa Router or handler isa StaticHandler then
+			route = new AppGlobRoute(path)
+		else if path.has_suffix("*") then
+			route = new AppGlobRoute(path)
+		else
+			route = new AppParamRoute(path)
+		end
+		handlers[route] = handler
+	end
+
+	redef fun handle(route, uri, req, res) do
+		if not route.match(uri) then return
+		for hroute, handler in handlers do
+			handler.handle(hroute, route.uri_root(uri), req, res)
+			if res.sent then break
+		end
+	end
+end
+
+# Popcorn application.
+#
+# The `App`  is the main point of the application.
+# It acts as a `Router` that holds the top level route handlers.
+#
+# Here an example to create a simple web app with Popcorn:
+#
+# ~~~
+# import popcorn
+#
+# class HelloHandler
+#	super Handler
+#
+#	redef fun get(req, res) do res.html "<h1>Hello World!</h1>"
+# end
+#
+# var app = new App
+# app.use("/", new HelloHandler)
+# # app.listen("localhost", 3000)
+# ~~~
+#
+# The Popcorn app listens on port 3000 for connections.
+# The app responds with "Hello World!" for request to the root URL (`/`) or **route**.
+# For every other path, it will respond with a **404 Not Found**.
+#
+# The `req` (request) and `res` (response) parameters are the same that nitcorn provides
+# so you can do anything else you would do in your route without Popcorn involved.
+#
+# Run the app with the following command:
+#
+# ~~~bash
+# nitc app.nit && ./app
+# ~~~
+#
+# Then, load [http://localhost:3000](http://localhost:3000) in a browser to see the output.
+class App
+	super Router
+end
+
+redef class HttpResponse
+
+	# Was this request sent by a handler?
+	var sent = false
+
+	private fun check_sent do
+		if sent then print "Warning: Headers already sent!"
```

Comment:
To be honest, I still not sure if what should be done here...

Should a handler that write the response terminate the response cycle?
Should I separate middlewares and default handlers, this part still in WIP until I develop more client to test it.